New Chain Integration + Minor changes to Chain Integration Overview (#725)
graphprotocol/docs: website/pages/en/new-chain-integration.mdx

---
title: New Chain Integration
---

Chains can bring subgraph support to their ecosystem by starting a new `graph-node` integration. Subgraphs are a powerful indexing tool opening a world of possibilities for developers. Graph Node already indexes data from the chains listed here. If you are interested in a new integration, there are 2 integration strategies:

1. **EVM JSON-RPC**
2. **Firehose**: All Firehose integration solutions include Substreams, a large-scale streaming engine based off Firehose with native `graph-node` support, allowing for parallelized transforms.

> > Note that while the recommended approach is to develop a new Firehose for all new chains, it is only required required for non-EVM chains.

Reviewer: ```suggestion
> Note that while the recommended approach is to develop a new Firehose for all new chains, it is only required for non-EVM chains.
```

Graph Node should be syncing the deployed subgraph if there are no errors. Give it time to sync, then send some GraphQL queries to the API endpoint printed in the logs.

## Substreams-powered Subgraphs

For StreamingFast-led Firehose/Substreams integrations, basic support for foundational Substreams modules (e.g. decoded transactions, logs and smart-contract events) and Substreams-powered subgraph codegen tools are included (check out [Injective](https://substreams.streamingfast.io/documentation/intro-getting-started/intro-injective/injective-first-sps) for an example).

There are two options to consume Substreams data through a subgraph:

- **Using Substreams triggers:** Consume from any Substreams module by importing the Protobuf model through a subgraph handler and move all your logic into a subgraph. This method creates the subgraph entities directly in the subgraph.
- **Using EntityChanges:** By writing more of the logic into Substreams, you can consume the module's output directly into `graph-node`. In `graph-node`, you can use the Substreams data to create your subgraph entities.

> It is really a matter of where you put your logic, in the subgraph or the Substreams. Keep in mind that having your logic in Substreams benefits from a parallelized model, whereas Substreams triggers will be linearly consumed. Consider the following example implementing a subgraph handler:

Reviewer: > Keep in mind that having your logic in Substreams benefits from a parallelized model, whereas Substreams triggers will be linearly consumed.

Just want to double check here, do we really want to say "Substreams" twice in this sentence or should one of them be "subgraphs"?

---
title: New Chain Integration
---

Chains can bring subgraph support to their ecosystem by starting a new `graph-node` integration. Subgraphs are a powerful indexing tool opening a world of possibilities for developers. Graph Node already indexes data from the chains listed here. If you are interested in a new integration, there are 2 integration strategies:

1. **EVM JSON-RPC**
2. **Firehose**: All Firehose integration solutions include Substreams, a large-scale streaming engine based off Firehose with native `graph-node` support, allowing for parallelized transforms.

> > Note that while the recommended approach is to develop a new Firehose for all new chains, it is a requirement for non-EVM chains.

Reviewer: Nitpick:
```suggestion
> Note that while the recommended approach is to develop a new Firehose for all new chains, it is only a requirement for non-EVM chains.
```

Reviewer: or even:
```suggestion
> Note that while the recommended approach is to develop a new Firehose for all new chains, it is not a required for EVM chains.
```

Reviewer: (basically, I feel like there's a tiny bit of ambiguity with the sentence's logic as it is)

## Integration Strategies

### 1. EVM JSON-RPC

If the blockchain is EVM equivalent and the client/node exposes the standard EVM JSON-RPC API, Graph Node should be able to index the new chain.

#### Testing an EVM JSON-RPC

> For Graph Node to be able to ingest data from an EVM chain, the RPC node must expose the following EVM JSON RPC methods:

Reviewer: ```suggestion
For Graph Node to be able to ingest data from an EVM chain, the RPC node must expose the following EVM JSON-RPC methods:
```

> NOTE: All integrations done by the StreamingFast team include maintenance for the Firehose replication protocol into the chain's codebase. StreamingFast tracks any changes and releases binaries when you change code and when StreamingFast changes code. This includes releasing Firehose/Substreams binaries for the protocol, maintaining Substreams modules for the block model of the chain, and releasing binaries for the blockchain node with instrumentation if need be.

#### Specific Firehose Instrumentation for EVM (`geth`) chains

> For the EVM chains, there exists a deeper level of data that can be achieved through the `geth` [live-tracer](https://github.com/ethereum/go-ethereum/releases/tag/v1.14.0), a collaboration between Go-Ethereum and StreamingFast, in building a high-throughput and rich transaction tracing system. The Live Tracer is the most comprehensive solution, resulting in [Extended](https://streamingfastio.medium.com/new-block-model-to-accelerate-chain-integration-9f65126e5425) block details. This enables new indexing paradigms, like pattern matching of events based on state changes, calls, parent call trees, or triggering of events based on changes to the actual variables in a smart contract.

Reviewer: ```suggestion
For EVM chains, there exists a deeper level of data that can be achieved through the `geth` [live-tracer](https://github.com/ethereum/go-ethereum/releases/tag/v1.14.0), a collaboration between Go-Ethereum and StreamingFast, in building a high-throughput and rich transaction tracing system. The Live Tracer is the most comprehensive solution, resulting in [Extended](https://streamingfastio.medium.com/new-block-model-to-accelerate-chain-integration-9f65126e5425) block details. This enables new indexing paradigms, like pattern matching of events based on state changes, calls, parent call trees, or triggering of events based on changes to the actual variables in a smart contract.
```

![Base block vs Extended block](/img/extended-vs-base-substreams-blocks.png)

> NOTE: This improvement upon the Firehose requires chains make use of the EVM engine `geth version 1.13.0` and up.

## EVM considerations - Difference between JSON-RPC & Firehose

While the JSON-RPC and Firehose are both suitable for subgraphs, a Firehose is always required for developers wanting to build with [Substreams](https://substreams.streamingfast.io). Supporting Substreams allows developers to build [Substreams-powered subgraphs](/cookbook/substreams-powered-subgraphs) for the new chain, and has the potential to improve the performance of your subgraphs. Additionally, Firehose — as a drop-in replacement for the JSON-RPC extraction layer of `graph-node` — reduces by 90% the number of RPC calls required for general indexing.

> - All those `getLogs` calls and roundtrips get replaced by a single stream arriving into the heart of `graph-node` and by running filters/transformations (Substreams) on the in-memory blocks; a single block model for all subgraphs it processes.

Reviewer: I don't understand this sentence. "and by running" is where I get confused – should it say "and running" or "by running"?

Author: No its additional to, like you can choose to do further filters and transformations - but I see confusion, I revised the sentence.

> NOTE: A Firehose-based integration for EVM chains will still require Indexers to run the chain's archive RPC node to properly index subgraphs. This is due to the Firehose's inability to provide smart contract state typically accessible by the `eth_call` RPC method. (It's worth reminding that `eth_calls` are not a good practice for developers)

## Graph Node Configuration

Configuring Graph Node is as easy as preparing your local environment. Once your local environment is set, you can test the integration by locally deploying a subgraph.

1. [Clone Graph Node](https://github.com/graphprotocol/graph-node)

> 2. Modify [this line](https://github.com/graphprotocol/graph-node/blob/master/docker/docker-compose.yml#L22) to include the new network name and the EVM JSON RPC compliant URL

Reviewer: ```suggestion
2. Modify [this line](https://github.com/graphprotocol/graph-node/blob/master/docker/docker-compose.yml#L22) to include the new network name and the EVM JSON-RPC compliant URL
```

   > Do not change the env var name itself. It must remain `ethereum` even if the network name is different.

3. Run an IPFS node or use the one used by The Graph: https://api.thegraph.com/ipfs/

> ### **Testing an EVM JSON-RPC by locally deploying a subgraph**

Reviewer: ```suggestion
### Testing an EVM JSON-RPC by locally deploying a subgraph
```

1. Install [graph-cli](https://github.com/graphprotocol/graph-cli)
2. Create a simple example subgraph. Some options are below:
   1. The pre-packed [Gravitar](https://github.com/graphprotocol/example-subgraph/tree/f89bdd4628efa4badae7367d4919b3f648083323) smart contract and subgraph is a good starting point
   2. Bootstrap a local subgraph from any existing smart contract or solidity dev environment [using Hardhat with a Graph plugin](https://github.com/graphprotocol/hardhat-graph)

> 3. Adapt the resulting `subgraph.yaml` by changing [`dataSources.network`](http://datasources.network/) to the same name previously passed on to Graph Node.

Reviewer: Did you mean to turn `dataSources.network` into a link? It doesn't seem to work...

There are two options to consume Substreams data through a subgraph:

- **Using Substreams triggers:** Consume from any Substreams module by importing the Protobuf model and move all your logic into the subgraph. This method creates the subgraph entities directly in the subgraph.
- **Using EntityChanges:** By writing more of the logic into Substreams, you can consume the module's output directly into `graph-node`. In `graph-node`, you can use the Substreams data to create your subgraph entities.

It is really a matter of where you put your logic, in the subgraph or the Substreams. Consider the following example using a subgraph handler:

```ts
export function handleTransactions(bytes: Uint8Array): void {
  let transactions = assembly.eth.transaction.v1.Transactions.decode(bytes.buffer).trasanctions // 1.
  if (transactions.length == 0) {
    log.info('No transactions found', [])
    return
  }

  for (let i = 0; i < transactions.length; i++) {
    // 2.
    let transaction = transactions[i]

    let entity = new Transaction(transaction.hash) // 3.
    entity.from = transaction.from
    entity.to = transaction.to
    entity.save()
  }
}
```

> The `handleTransactions` function is a subgraph handler that receives the raw Substreams bytes as parameter and decodes them into a “Transactions” object. Then, for every transaction, a new subgraph entity is created. For more information about Substreams triggers, visit the [StreamingFast documentation](https://substreams.streamingfast.io/documentation/consume/subgraph/triggers) or check out community modules at [substreams.dev](https://substreams.dev/).

Reviewer: ```suggestion
The `handleTransactions` function is a subgraph handler that receives the raw Substreams bytes as parameter and decodes them into a `Transactions` object. Then, for every transaction, a new subgraph entity is created. For more information about Substreams triggers, visit the [StreamingFast documentation](https://substreams.streamingfast.io/documentation/consume/subgraph/triggers) or check out community modules at [substreams.dev](https://substreams.dev/).
```

> > **NOTE:** Keep in mind that having your logic in Substreams benefits from a parallelized model, whereas Substreams triggers will be linearly consumed.

Reviewer: ```suggestion
> NOTE: Keep in mind that having your logic in Substreams benefits from a parallelized model, whereas Substreams triggers will be linearly consumed.
```
Just to be consistent with the other notes on page. Actually, I don't think "NOTE:" is even necessary, but whatever. :P

---

> title: New Chain Integration

Reviewer: I noticed that the nav still calls this page "Integrating New Networks", and that's because it's been defined in `website/pages/en/_meta.js`. You should replace
```
'new-chain-integration': 'Integrating New Networks',
```
with
```
'new-chain-integration': '',
```

---

> Chains can bring subgraph support to their ecosystem by starting a new graph-node integration. Subgraphs are a powerful indexing tool opening a world of possibilities for developers. Graph Node already indexes data from the chains listed here. If you are interested in a new integration, there are 2 integration strategies:

Reviewer: ```suggestion
Chains can bring subgraph support to their ecosystem by starting a new `graph-node` integration. Subgraphs are a powerful indexing tool opening a world of possibilities for developers. Graph Node already indexes data from the chains listed here. If you are interested in a new integration, there are 2 integration strategies:
```

1. **EVM JSON-RPC**

> 2. **Firehose**: All Firehose integration solutions include Substreams, a large scale streaming engine based off Firehose with native `graph-node` support, allowing for parallelized transforms.

Reviewer: ```suggestion
2. **Firehose**: All Firehose integration solutions include Substreams, a large-scale streaming engine based off Firehose with native `graph-node` support, allowing for parallelized transforms.
```

> > Note that while the recommended approach is to developing a new Firehose for all new chains, it is a requirement for non-EVM chains.

Reviewer: ```suggestion
> Note that while the recommended approach is to develop a new Firehose for all new chains, it is a requirement for non-EVM chains.
```

## Integration Strategies

### 1. EVM JSON-RPC

If the blockchain is EVM equivalent and the client/node exposes the standard EVM JSON-RPC API, Graph Node should be able to index the new chain.

#### Testing an EVM JSON-RPC

For Graph Node to be able to ingest data from an EVM chain, the RPC node must expose the following EVM JSON RPC methods:

- `eth_getLogs`
- `eth_call` (for historical blocks, with EIP-1898 - requires archive node)
- `eth_getBlockByNumber`
- `eth_getBlockByHash`
- `net_version`
- `eth_getTransactionReceipt`, in a JSON-RPC batch request

> - `*trace_filter*` *(optionally required for Graph Node to support call handlers)*

Reviewer: ```suggestion
- `trace_filter` *(optionally required for Graph Node to support call handlers)*
```

### 2. Firehose Integration

> [Firehose](https://firehose.streamingfast.io/firehose-setup/overview) is a next generation extraction layer. It collects history in flat files and streams in real-time. Firehose technology replaces those polling API calls with a stream of data utilizing a push model that sends data to the indexing node faster. This helps increase the speed of syncing and indexing.

Reviewer: ```suggestion
[Firehose](https://firehose.streamingfast.io/firehose-setup/overview) is a next-generation extraction layer. It collects history in flat files and streams in real time. Firehose technology replaces those polling API calls with a stream of data utilizing a push model that sends data to the indexing node faster. This helps increase the speed of syncing and indexing.
```

> The primary method to integrate the Firehose into chains is to use an RPC polling strategy. Our polling algorithm will predict when a new block will arrive and increase the rate at which it checks for a new block near that time, making it a very low-latency and efficient solution. For help with the integration and maintenance of the Firehose contact the StreamingFast team. New chains and their integrators will appreciate the [fork awareness](https://substreams.streamingfast.io/documentation/consume/reliability-guarantees) and massive parallelized indexing capabilities that Firehose and Substreams brings to their ecosystem.

Reviewer: ```suggestion
The primary method to integrate the Firehose into chains is to use an RPC polling strategy. Our polling algorithm will predict when a new block will arrive and increase the rate at which it checks for a new block near that time, making it a very low-latency and efficient solution. For help with the integration and maintenance of the Firehose, contact the StreamingFast team. New chains and their integrators will appreciate the [fork awareness](https://substreams.streamingfast.io/documentation/consume/reliability-guarantees) and massive parallelized indexing capabilities that Firehose and Substreams brings to their ecosystem.
```
(also, should "contact the StreamingFast team" be a link?)

Reviewer 2: good call. @Giuliano-1 , you may want to add a link/e-mail/telegram id here, right?

Reviewer 2: Nevermind. I now see it is there.

> NOTE: All integrations done by the StreamingFast team include maintenance for the Firehose replication protocol into the chain's codebase. StreamingFast tracks any changes and releases binaries when you change code and when StreamingFast changes code. This includes releasing Firehose/Substreams binaries for the protocol, maintaining Substreams modules for the block model of the chain, and releasing binaries for the blockchain node with instrumentation if need be.

#### Specific Firehose Instrumentation for EVM (`geth`) chains

> For the EVM chains, there exists a deeper level of data that can be achieved through the `geth` [live-tracer](https://github.com/ethereum/go-ethereum/releases/tag/v1.14.0), a collaboration between Go-Ethereum and StreamingFast, in building a high-throughput and rich transaction tracing system. The Live Tracer is the most comprehensive solution, resulting in [EXTENDED](https://streamingfastio.medium.com/new-block-model-to-accelerate-chain-integration-9f65126e5425) block details. This enables new indexing paradigms, like pattern matching of events based on state changes, calls, parent call trees, or triggering of events based on changes to the actual variables in a smart contract.

Reviewer: Just curious, why is "EXTENDED" uppercase?

Author: You're right, it's not necessary to uppercase each character

![Base block vs Extended block](/img/extended-vs-base-substreams-blocks.png)

> NOTE: This improvement upon the Firehose requires chains make use of the EVM engine `geth version 1.13.0` and up.

## EVM considerations - Difference between JSON-RPC & Firehose

> While the JSON-RPC and Firehose are both suitable for subgraphs, a Firehose is always required for developers wanting to build with [Substreams](https://substreams.streamingfast.io). Supporting Substreams allows developers to build [Substreams-powered subgraphs](https://thegraph.com/docs/en/cookbook/substreams-powered-subgraphs) for the new chain, and has the potential to improve the performance of your subgraphs. Additionally, Firehose — as a drop-in replacement for the JSON-RPC extraction layer of `graph-node` — reduces by 90% the number of RPC calls required for general indexing.

Reviewer: ```suggestion
While the JSON-RPC and Firehose are both suitable for subgraphs, a Firehose is always required for developers wanting to build with [Substreams](https://substreams.streamingfast.io). Supporting Substreams allows developers to build [Substreams-powered subgraphs](/cookbook/substreams-powered-subgraphs) for the new chain, and has the potential to improve the performance of your subgraphs. Additionally, Firehose — as a drop-in replacement for the JSON-RPC extraction layer of `graph-node` — reduces by 90% the number of RPC calls required for general indexing.
```

> - All those `getLogs` call and roundtrips get replaced by a single stream arriving into the heart of `graph-node` and by running filters on the in-memory Firehose Block; a _single_ block for all subgraphs it processes.

Reviewer: ```suggestion
- All those `getLogs` calls and roundtrips get replaced by a single stream arriving into the heart of `graph-node` and by running filters on the in-memory Firehose Block; a _single_ block for all subgraphs it processes.
```
(also, I'm not sure what is "running filters")

Author: it's referring to a Substreams block as in a non-full block anymore... considering to rephrase it

> NOTE: A Firehose-based integration for EVM chains will still require Indexers to run the chain's archive RPC node to properly index subgraphs. This is due to the Firehose's inability to provide smart contract state typically accessible by the `eth_call` RPC method. (It's worth reminding that `eth_calls` are not a good practice for developers)

## Graph Node Configuration

Configuring Graph Node is as easy as preparing your local environment. Once your local environment is set, you can test the integration by locally deploying a subgraph.

1. [Clone Graph Node](https://github.com/graphprotocol/graph-node)
2. Modify [this line](https://github.com/graphprotocol/graph-node/blob/master/docker/docker-compose.yml#L22) to include the new network name and the EVM JSON RPC compliant URL

>    > Do not change the env var name itself. It must remain ethereum even if the network name is different.

Reviewer: ```suggestion
   > Do not change the env var name itself. It must remain `ethereum` even if the network name is different.
```

3. Run an IPFS node or use the one used by The Graph: https://api.thegraph.com/ipfs/

> <details>

Reviewer: Erm, I'm guessing this whole section written in HTML does not look like you expect. 😅

![CleanShot 2024-07-11 at 01 40 31@2x](https://github.com/graphprotocol/docs/assets/1059139/274cae7e-fa22-489e-baee-64c6b8b58576)

Does it really have to be an accordion? If so, I'm sure there's a way we can do it properly in MDX (probably by providing the GDS `Accordion` component), but for now it would probably be easier just to turn "Testing an EVM JSON-RPC by locally deploying a subgraph" into a heading, the `<ol>` into a Markdown numbered list, the `<code>`s into backtick-wrapped strings, etc. What do you think?

Reviewer 2: @Giuliano-1 , it may be best to return to the idea of having this as a dedicated sub-page instead.

Author: The changes made for `lint` and `prettier` did not result well but it should work with the embedded HTML

Author: @benface this is what I see when I "git pull" this code (edited for `lint` and `prettier`) and test upload it to my personal github repo (or when navigating here: https://github.com/graphprotocol/docs/blob/giuliano/docrevision/website/pages/en/new-chain-integration.mdx)... it looks normal, can you give me an idea as to why it appears different on your end? 

The reason I want the `Accordion` to work is for a richer reading experience!

<img width="1070" alt="Screenshot 2024-07-22 at 11 07 48 PM" src="https://github.com/user-attachments/assets/9f22839d-a614-432d-858d-491558313b10">

Reviewer: @Giuliano-1 – Looks like you're previewing the file in the GitHub UI, not in the actual Graph docs app. Do you have the repo cloned locally? If so, you can see what I see by running `pnpm dev` and going to `http://localhost:3000/en/new-chain-integration/` in your browser.

Reviewer: (See the [`README`](https://github.com/graphprotocol/docs/blob/main/README.md) for more detailed instructions)

        Gravitar
      </a>
      smart contract and subgraph is a good starting point
      <br />
    </li>
    <li>
      Bootstrap a local subgraph from any existing smart contract or solidity dev environment 
      <a href="https://github.com/graphprotocol/hardhat-graph">using Hardhat with a Graph plugin</a>
      <br />
    </li>
  </ol>
  <li>
    Adapt the resulting <code>subgraph.yaml</code> by changing 
    <a href="http://datasources.network/">
      <code>dataSources.network</code>
    </a>
     to the same name previously passed on to Graph Node.
    <br />
  </li>
  <li>
    Create your subgraph in Graph Node: <code>graph create $SUBGRAPH_NAME --node $GRAPH_NODE_ENDPOINT</code>
    <br />
  </li>
  <li>
    Publish your subgraph to Graph Node: 
    <code>graph deploy $SUBGRAPH_NAME --ipfs $IPFS_ENDPOINT --node $GRAPH_NODE_ENDPOINT</code>
    <br />
  </li>
</ol>

Graph Node should be syncing the deployed subgraph if there are no errors. Give it time to sync, then send some GraphQL queries to the API endpoint printed in the logs.<br/>

</details>
    
## Substreams-powered Subgraphs

> For StreamingFast led Firehose/Substreams integrations, basic support for foundational Substreams modules (e.g. decoded transactions, logs and smart-contract events) and Substreams-powered-Subgraph codegen tools are included (check out [Injective](https://substreams.streamingfast.io/documentation/intro-getting-started/intro-injective/injective-first-sps) for an example).

Reviewer: ```suggestion
For StreamingFast-led Firehose/Substreams integrations, basic support for foundational Substreams modules (e.g. decoded transactions, logs and smart-contract events) and Substreams-powered subgraph codegen tools are included (check out [Injective](https://substreams.streamingfast.io/documentation/intro-getting-started/intro-injective/injective-first-sps) for an example).
```

There are two options to consume Substreams data through a subgraph:

> - **Using Substreams triggers:** Consume from any Substreams module by importing the Protobuf model and move all your logic into the subgraph. This method creates the subgraph entities directly in the subgraph

Reviewer: ```suggestion
- **Using Substreams triggers:** Consume from any Substreams module by importing the Protobuf model and move all your logic into the subgraph. This method creates the subgraph entities directly in the subgraph.
```

> - **Using EntityChanges:** By writing more of the logic into Substreams, you can consume the module-output directly into `graph-node`. In `graph-node`, you can use the Substreams data to create your subgraph entities.

Reviewer: ```suggestion
- **Using EntityChanges:** By writing more of the logic into Substreams, you can consume the module's output directly into `graph-node`. In `graph-node`, you can use the Substreams data to create your subgraph entities.
```

It is really a matter of where you put your logic, in the subgraph or the Substreams. Consider the following example using a subgraph handler:

> ```

Reviewer 3: ```suggestion
```ts
```

export function handleTransactions(bytes: Uint8Array): void {
    let transactions = assembly.eth.transaction.v1.Transactions.decode(bytes.buffer).trasanctions; // 1.
    if (transactions.length == 0) {
        log.info("No transactions found", []);
        return;
    }

    for (let i = 0; i < transactions.length; i++) { // 2.
        let transaction = transactions[i];

        let entity = new Transaction(transaction.hash); // 3.
        entity.from = transaction.from;
        entity.to = transaction.to;
        entity.save();
    }
}
```

> The “handleTransactions” function is a subgraph handler that receives the raw Substreams bytes as parameter and decodes them into a “Transactions” object. Then, for every transaction, a new subgraph entity is created. For more information about Substreams triggers, visit the [StreamingFast documentation](https://substreams.streamingfast.io/documentation/consume/subgraph/triggers) or checkout community modules at [substreams.dev](http://substreams.dev).

Reviewer: ```suggestion
The `handleTransactions` function is a subgraph handler that receives the raw Substreams bytes as parameter and decodes them into a “Transactions” object. Then, for every transaction, a new subgraph entity is created. For more information about Substreams triggers, visit the [StreamingFast documentation](https://substreams.streamingfast.io/documentation/consume/subgraph/triggers) or check out community modules at [substreams.dev](https://substreams.dev).
```

> > **NOTE:** Keep in mind that having your logic in Substreams benefits from a parallelized model, whereas, Substreams triggers will be linearly consumed.

Reviewer: ```suggestion
> **NOTE:** Keep in mind that having your logic in Substreams benefits from a parallelized model, whereas Substreams triggers will be linearly consumed.
```
